core/domain/caching_services_test.py

```diff
@@ -381,3 +393,199 @@ def test_all_namespace_strings_does_not_contain_memcache_delimiter(self):
         # dictionary.
         for namespace in caching_services.SERIALIZATION_FUNCTIONS:
             self.assertNotIn(caching_services.MEMCACHE_KEY_DELIMITER, namespace)
+
+    def test_unicode_characters_are_set_and_get_correctly_in_default_namespace(
+            self):
+        # Test to make sure that unicode characters are still inserted and
+        # deleted correctly from the cache.
+        key_value_mapping = {
+            'a': '%#$', 'b': '\t',
+            'c': '😃😄'
+        }
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_DEFAULT, None,
+                ['a', 'b', 'c']),
+            {})
+        cache_strings_response = caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_DEFAULT, None, key_value_mapping)
+        self.assertTrue(cache_strings_response)
+
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_DEFAULT, None,
+                ['a', 'b', 'c']),
+            {
+                'a': '%#$', 'b': '\t',
+                'c': '😃😄'
+            })
+
+    def test_objects_with_unicode_characters_are_set_and_get_correctly(
+            self):
+        # Test to make sure that a default exploration initialized with unicode
+        # characters is get and set to the cache without errors.
+        exploration_id = 'id'
+
+        self.assertEqual(
+            caching_services.get_multi(
+                caching_services.CACHE_NAMESPACE_EXPLORATION,
+                '0',
+                [exploration_id]),
+            {})
+
+        default_exploration = (
+            exp_domain.Exploration.create_default_exploration(
+                exploration_id, title='A title',
+                category='A category 😍',
+                objective='Objective',
+                language_code=['en', 'zh']
+            ))
+
+        caching_services.set_multi(
+            caching_services.CACHE_NAMESPACE_EXPLORATION,
+            '0',
+            {
+                exploration_id: default_exploration
+            })
+
+        exp_ids_to_explorations = caching_services.get_multi(
+            caching_services.CACHE_NAMESPACE_EXPLORATION,
+            '0', [exploration_id])
+
+        self.assertEqual(
+            default_exploration.to_dict(),
+            exp_ids_to_explorations[exploration_id].to_dict())
+
+        # Test to make sure that a default collection initialized with unicode
```

Comment:
Put separate tests in separate test methods. See guidelines [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

Reply:
Ah, I see. I know about the guidelines and was debating about whether to keep it in one or separate functions but I figured that testing that each of the objects contains unicode characters satisfies the umbrella of one function, and it made the code less messy.

I'll change it and keep everything in separate tests, even if they have some tangential similarities from now on.